visualizer/src/Canvas/Labeled/LabeledCanvas.js

```diff
@@ -1,104 +1,66 @@
-  const labeledImage = useSelector(feature, (state) => state.context.labeledImage);
+  const colormap = useSelector(feature, (state) => state.context.colormap);
   let labeledArray = useSelector(feature, (state) => state.context.labeledArray);
   if (!labeledArray) {
-    labeledArray = Array(sh).fill(Array(sw).fill(0));
+    labeledArray = Array(height * width).fill(0);
   }
 
   const select = useSelect();
   const foreground = useSelector(select, (state) => state.context.foreground);
-  const background = useSelector(select, (state) => state.context.background);
 
+  const kernel = useRef();
   const canvasRef = useRef();
-  const ctx = useRef();
-  const hiddenCanvasRef = useRef();
-  const hiddenCtx = useRef();
 
   useEffect(() => {
-    ctx.current = canvasRef.current.getContext('2d');
-    ctx.current.imageSmoothingEnabled = false;
+    const canvas = canvasRef.current;
+    const gl = canvas.getContext('webgl2', { premultipliedAlpha: false });
+    const gpu = new GPU({
+      canvas,
+      gl,
+    });
+    kernel.current = gpu
+      .createKernel(function (labels, colormap, foreground, highlight, highlightColor, opacity) {
+        const n = this.thread.x + this.constants.w * (this.constants.h - 1 - this.thread.y);
+        const label = labels[n];
+        if (highlight && label === foreground && foreground !== 0) {
+          const [r, g, b] = highlightColor;
+          this.color(r / 255, g / 255, b / 255, opacity);
+        } else if (label !== 0) {
+          const [r, g, b] = colormap[label - 1];
+          this.color(r / 255, g / 255, b / 255, opacity);
+        } else {
+          this.color(0, 0, 0, opacity);
+        }
+      })
+      .setConstants({ w: width, h: height })
+      .setOutput([width, height])
+      .setGraphical(true)
+      .setDynamicArguments(true);
   }, [width, height]);
 
   useEffect(() => {
-    hiddenCtx.current = hiddenCanvasRef.current.getContext('2d');
-  }, [sw, sh]);
-
-  useEffect(() => {
-    hiddenCtx.current.drawImage(labeledImage, 0, 0);
-    let data = hiddenCtx.current.getImageData(0, 0, sw, sh);
-    if (highlight && foreground !== 0) {
-      const red = [255, 0, 0, 255];
-      highlightImageData(data, labeledArray, foreground, red);
-    }
-    opacityImageData(data, opacity);
-    hiddenCtx.current.putImageData(data, 0, 0);
-  }, [labeledImage, labeledArray, foreground, highlight, opacity, sh, sw]);
-
-  useEffect(() => {
-    ctx.current.save();
-    ctx.current.clearRect(0, 0, width, height);
-    ctx.current.drawImage(
-      hiddenCanvasRef.current,
-      sx,
-      sy,
-      sw / zoom,
-      sh / zoom,
-      0,
-      0,
-      width,
-      height
-    );
-    ctx.current.restore();
-  }, [
-    labeledImage,
-    labeledArray,
-    foreground,
-    highlight,
-    opacity,
-    sw,
-    sh,
-    sx,
-    sy,
-    zoom,
-    width,
-    height,
-  ]);
+    kernel.current(labeledArray, colormap, foreground, highlight, highlightColor, opacity);
```

Comment:
Could you please add some brief comments explaining what the effect is supposed to do?

Reply:
The `kernel.current(...)` line rerenders the canvas powered by GPU.js for the current component. The next line `setCanvases(...)` rerenders the parent canvas with the updated canvas for this component. I added these comments to these effects in each Canvas component that uses it.